model/src/lib.rs

```diff
@@ -52,26 +58,97 @@ pub struct OrderCreation {
 }
 
 impl OrderCreation {
+    pub const ORDER_TYPE_HASH: [u8; 32] =
+        hex!("b71968fcf5e55b9c3370f2809d4078a4695be79dfa43e5aa1f2baa0a9b84f186");
+
     pub fn token_pair(&self) -> Option<TokenPair> {
         TokenPair::new(self.buy_token, self.sell_token)
     }
-    pub fn order_owner(&self) -> H160 {
-        // dummy implementation, waiting for Valentins PR
-        H160::zero()
+
+    pub fn order_digest(&self) -> [u8; 32] {
+        let mut hash_data = [0u8; 320];
+        hash_data[0..32].copy_from_slice(&Self::ORDER_TYPE_HASH);
+        // Some slots are not assigned (stay 0) because all values are extended to 256 bits.
+        hash_data[44..64].copy_from_slice(self.sell_token.as_fixed_bytes());
+        hash_data[76..96].copy_from_slice(self.buy_token.as_fixed_bytes());
+        self.sell_amount.to_big_endian(&mut hash_data[96..128]);
+        self.buy_amount.to_big_endian(&mut hash_data[128..160]);
+        hash_data[188..192].copy_from_slice(&self.valid_to.to_be_bytes());
+        hash_data[220..224].copy_from_slice(&self.app_data.to_be_bytes());
+        self.fee_amount.to_big_endian(&mut hash_data[224..256]);
+        hash_data[287] = match self.order_kind {
+            OrderKind::Sell => 0,
+            OrderKind::Buy => 1,
+        };
+        hash_data[319] = self.partially_fillable as u8;
+        signing::keccak256(&hash_data)
+    }
+
+    pub fn signing_digest_typed_data(&self, domain_separator: &[u8; 32]) -> [u8; 32] {
```

Comment:
can these be private? I would think only `validate_signature` and `uid` need to be public?

Reply:
Made private